SCons/Tool/ninja.py

```diff
@@ -0,0 +1,1570 @@
+            ninja_syntax_file = env.File(ninja_syntax_file).get_abspath()
+        ninja_syntax_mod_dir = os.path.dirname(ninja_syntax_file)
+        sys.path.append(ninja_syntax_mod_dir)
+        ninja_syntax_mod_name = os.path.basename(ninja_syntax_file).replace(".py", "")
+        ninja_syntax = importlib.import_module(ninja_syntax_mod_name)
+    else:
+        ninja_syntax = importlib.import_module(".ninja_syntax", package='ninja')
+        
+    if NINJA_STATE is None:
+        NINJA_STATE = NinjaState(env, ninja_file[0], ninja_syntax.Writer)
+        NINJA_STATE.ninja_bin_path = os.path.abspath(os.path.join(ninja_syntax.__file__, os.pardir, 'data', 'bin'))
+    # Here we will force every builder to use an emitter which makes the ninja
+    # file depend on it's target. This forces the ninja file to the bottom of
+    # the DAG which is required so that we walk every target, and therefore add
+    # it to the global NINJA_STATE, before we try to write the ninja file.
+    def ninja_file_depends_on_all(target, source, env):
+        if not any("conftest" in str(t) for t in target):
+            env.Depends(ninja_file, target)
+        return target, source
+
+    # The "Alias Builder" isn't in the BUILDERS map so we have to
+    # modify it directly.
+    SCons.Environment.AliasBuilder.emitter = ninja_file_depends_on_all
+
+    for _, builder in env["BUILDERS"].items():
+        try:
+            emitter = builder.emitter
+            if emitter is not None:
+                builder.emitter = SCons.Builder.ListEmitter(
+                    [emitter, ninja_file_depends_on_all]
+                )
+            else:
+                builder.emitter = ninja_file_depends_on_all
+        # Users can inject whatever they want into the BUILDERS
+        # dictionary so if the thing doesn't have an emitter we'll
+        # just ignore it.
+        except AttributeError:
+            pass
+
+    # We will subvert the normal Command to make sure all targets generated
+    # from commands will be linked to the ninja file
+    SconsCommand = SCons.Environment.Environment.Command
+
+    def NinjaCommand(self, target, source, action, **kw):
+        targets = SconsCommand(env, target, source, action, **kw)
+        env.Depends(ninja_file, targets)
+        return targets
+    
+    SCons.Environment.Environment.Command = NinjaCommand
+
+    # Here we monkey patch the Task.execute method to not do a bunch of
+    # unnecessary work. If a build is a regular builder (i.e not a conftest and
+    # not our own Ninja builder) then we add it to the NINJA_STATE. Otherwise we
+    # build it like normal. This skips all of the caching work that this method
+    # would normally do since we aren't pulling any of these targets from the
+    # cache.
+    #
+    # In the future we may be able to use this to actually cache the build.ninja
+    # file once we have the upstream support for referencing SConscripts as File
+    # nodes.
+    def ninja_execute(self):
```

Comment:
I think this could re-use more of SCons:

```
+
+    base_execute = SCons.Taskmaster.Task.execute
     def ninja_execute(self):
         global NINJA_STATE

         target = self.targets[0]
         target_name = str(target)
-        if target_name != ninja_file_name and "conftest" not in target_name:
-            NINJA_STATE.add_build(target)
-        else:
-            target.build()

+        if target_name == ninja_file_name or "conftest" in target_name:
+            base_execute(self)
+        else:
+            NINJA_STATE.add_build(target)

```